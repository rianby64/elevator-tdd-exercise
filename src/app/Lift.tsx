
import * as React from 'react';
import { LiftPanel } from './LiftPanel';

export enum Direction {
    Up = "UP",
    Down = "DOWN",
}

interface LiftProps {
    direction?: Direction;
    floors: number;
    currentLevel?: number;
    onCallLift?: (i: number) => void;
}
interface DefaultProps {
    currentLevel: number;
}

interface LiftState {
    highlightedButtons: Set<number>;
}

export class Lift extends React.Component<LiftProps & DefaultProps> {

    public static defaultProps: DefaultProps = {
        currentLevel: 1,
    }

    public state: LiftState = {
        highlightedButtons: new Set([]),
    }

    private callLift = (i: number): () => void => {
        return (): void => {
            this.setState((state: LiftState): LiftState => {
                const highlightedButtons = new Set([...state.highlightedButtons, i]);
                return {...state, highlightedButtons: highlightedButtons};
            });
            if (this.props.onCallLift) {
                this.props.onCallLift(i);
            }
        }
    }

    public componentDidUpdate(): void {
        if (this.state.highlightedButtons.has(this.props.currentLevel)) {
            this.setState((state: LiftState): LiftState => {
                const highlightedButtons = new Set(state.highlightedButtons)
                highlightedButtons.delete(this.props.currentLevel);
                return {...state, highlightedButtons: highlightedButtons}
            });
        }
    }

    public render(): JSX.Element {
        return (
            <div>
                {this.props.direction ?
                    <span>Going {this.props.direction}</span> : null}
                <LiftPanel
                    callLift={this.callLift}
                    highlightedButtons={this.state.highlightedButtons}
                    floors={this.props.floors}
                    currentLevel={this.props.currentLevel} />
            </div>
        );
    }
}
