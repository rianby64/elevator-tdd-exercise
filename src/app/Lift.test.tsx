
import * as React from 'react';
import { mount } from 'enzyme';
import { Lift, Direction } from './Lift';

test('Lift highlighs the currentLevel', (): void => {
    const el = (<Lift currentLevel={3} floors={5} />);
    const sh = mount(el);
    expect(sh).toMatchSnapshot();
});

test('Lift has the state highlightedButtons', (): void => {
    const el = (<Lift currentLevel={3} floors={5} />);
    const sh = mount(el);
    expect(sh.state()).toHaveProperty('highlightedButtons');
});

test('Lift highlighs the currentLevel and shows the direction', (): void => {
    const el = (<Lift currentLevel={3} floors={5} direction={Direction.Down} />);
    const sh = mount(el);
    expect(sh).toMatchSnapshot();
});

test('Lift highlighs a pressed button', (): void => {
    const el = (<Lift currentLevel={3} floors={5} />);
    const sh = mount(el);
    sh.find('LiftPanel ul').findWhere(el => el.key() === '2').find('button').simulate('click');
    expect(sh).toMatchSnapshot();
});
