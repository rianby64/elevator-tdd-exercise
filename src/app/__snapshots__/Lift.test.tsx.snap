// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lift highlighs the currentLevel 1`] = `
<Lift
  currentLevel={3}
  floors={5}
>
  <div>
    <LiftPanel
      currentLevel={3}
      floors={5}
      highlightButton={[Function]}
      highlightedButtons={Set {}}
    >
      <div>
        <span>
          Lift's panel
        </span>
        <ul>
          <li
            key="4"
          >
            <button
              onClick={[Function]}
            >
              5
            </button>
          </li>
          <li
            key="3"
          >
            <button
              onClick={[Function]}
            >
              4
            </button>
          </li>
          <li
            className="active"
            key="2"
          >
            <button
              onClick={[Function]}
            >
              3
            </button>
          </li>
          <li
            key="1"
          >
            <button
              onClick={[Function]}
            >
              2
            </button>
          </li>
          <li
            key="0"
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </li>
        </ul>
      </div>
    </LiftPanel>
  </div>
</Lift>
`;

exports[`Lift highlighs the currentLevel and shows the direction 1`] = `
<Lift
  currentLevel={3}
  direction="DOWN"
  floors={5}
>
  <div>
    <span>
      Going 
      DOWN
    </span>
    <LiftPanel
      currentLevel={3}
      floors={5}
      highlightButton={[Function]}
      highlightedButtons={Set {}}
    >
      <div>
        <span>
          Lift's panel
        </span>
        <ul>
          <li
            key="4"
          >
            <button
              onClick={[Function]}
            >
              5
            </button>
          </li>
          <li
            key="3"
          >
            <button
              onClick={[Function]}
            >
              4
            </button>
          </li>
          <li
            className="active"
            key="2"
          >
            <button
              onClick={[Function]}
            >
              3
            </button>
          </li>
          <li
            key="1"
          >
            <button
              onClick={[Function]}
            >
              2
            </button>
          </li>
          <li
            key="0"
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </li>
        </ul>
      </div>
    </LiftPanel>
  </div>
</Lift>
`;
